fdbclient/admin_client.go

```diff
@@ -706,3 +709,116 @@ func (client *cliAdminClient) getTimeout() time.Duration {
 
 	return client.timeout
 }
+
+// GetProcessesUnderMaintenance will return all process groups that are currently stored to be under maintenance.
+// The result is a map with the process group ID as key and the start of the maintenance as value.
+func (client *cliAdminClient) GetProcessesUnderMaintenance() (map[fdbv1beta2.ProcessGroupID]int64, error) {
+	db, err := getFDBDatabase(client.Cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	maintenancePrefix := client.Cluster.GetMaintenancePrefix() + "/"
+
+	maintenanceProcesses, err := db.Transact(func(tr fdb.Transaction) (interface{}, error) {
+		err := tr.Options().SetReadSystemKeys()
+		if err != nil {
+			return nil, err
+		}
+
+		keyRange, err := fdb.PrefixRange([]byte(client.Cluster.GetMaintenancePrefix()))
+		if err != nil {
+			return nil, err
+		}
+
+		results := tr.GetRange(keyRange, fdb.RangeOptions{}).GetSliceOrPanic()
+		upgrades := make(map[fdbv1beta2.ProcessGroupID]int64, len(results))
+		for _, result := range results {
+			var timestamp int64
+			parseErr := binary.Read(bytes.NewBuffer(result.Value), binary.LittleEndian, &timestamp)
+			if parseErr != nil {
+				client.log.Error(parseErr, "could not parse timestamp", "byteValue", string(result.Value))
+				// TODO anything better to do here?
```

Comment:
A user adding a wrong value could lead to an issue here or some automation adding the wrong value here. The returned values are not used to consider "what is safe to take down" if something is in this list, we have to assume it was already taken down. The information in this key-range if only used to identify when the maintenance mode should be reseted by the operator, e.g. once all processes in the key-range have been restarted after the added maintenance time.

A single process added with a bad value could block all further actions if we return here. The risk would be that the process that was added with a wrong value will be ignored in the check and the operator might reset the maintenance mode a bit earlier than it should be. This could lead to some data movement, but in general the risk is pretty low and I would rather favour that the operator is able to proceed.

Reply:
that makes much more sense, thank you for helping me understand!